content/en/BestPractices/prevent-xml-external-entities-attack.md

```diff
@@ -0,0 +1,292 @@
+
+```
+
+package com.mkyong.xml.sax;
+
+import com.mkyong.xml.sax.handler.PrintAllHandlerSax;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.parsers.SAXParser;
+import javax.xml.parsers.SAXParserFactory;
+import java.io.IOException;
+
+public class ReadXmlSaxParserXXE {
+
+  private static final String FILENAME = "src/main/resources/staff.xml";
+
+  public static void main(String[] args) {
+
+      SAXParserFactory factory = SAXParserFactory.newInstance();
+
+      try {
+
+          // XXE attack
+          SAXParser saxParser = factory.newSAXParser();
+
+          PrintAllHandlerSax handler = new PrintAllHandlerSax();
+
+          saxParser.parse(FILENAME, handler);
+
+      } catch (ParserConfigurationException | SAXException | IOException e) {
+          e.printStackTrace();
+      }
+
+  }
+
+}
+
+```
+
+An attacker can craft a payload like the one below to execute XXE in the above vulnerable code, which will return the contents of the `/etc/passwd` file. 
+
+```
+ <?xml version="1.0" encoding="UTF-8"?>
+  <!DOCTYPE foo [ <!ENTITY xxe SYSTEM "file:///etc/passwd"> ]>
+  <company>
+      <staffId>&xxe;</staffId>
+  </company>
+```
+
+However, this could be prevented by using `setFeature` to disable the DOCTYPE declaration and remediate the XXE attack, as shown in the below code example: 
+
+```
+  SAXParserFactory factory = SAXParserFactory.newInstance();
+
+  try {
+
+      // https://rules.sonarsource.com/java/RSPEC-2755
+      // prevent XXE, completely disable DOCTYPE declaration:
+      factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+
+      SAXParser saxParser = factory.newSAXParser();
+
+      PrintAllHandlerSax handler = new PrintAllHandlerSax();
+
+      saxParser.parse(FILENAME, handler);
+
+  } catch (ParserConfigurationException | SAXException | IOException e) {
+      e.printStackTrace();
+  }
+```
+
+You may refer to different Sonar Source Rules to understand the XXE remediation for famous XML parsers using rules such as https://rules.sonarsource.com/java/RSPEC-2755.
+
+
+#### Denial of Service - XML Billion Laugh Attack 
+
+XML parsers are prone to denial of service attacks when a malicious XML document containing the same large entity is repeated repeatedly, causing an infinite processing loop. In addition, due to the lack of restrictions in place, such as a limit on the number of entity expansions, the XML processor can consume a heavy amount of memory and time during the parsing resulting in a Denial of Service attack. 
+
+For example, in the below code example, when the `XMLConstants.FEATURE_SECURE_PROCESSING` feature is set to `false`, it could allow an attacker to perform an XML Billion Laugh Attack: 
+
+```
```

Comment:
```suggestion
To learn more about remediating XXE attacks for popular XML parsers, refer to resources such as [Sonar Source Rules](https://rules.sonarsource.com/java/RSPEC-2755).

### Denial of Service: XML Billion Laughs Attack

XML parsers are prone to Denial of Service attacks when a malicious XML document containing a large entity is repeated, causing an infinite processing loop. If there are no restrictions such as a limited number of entities, an XML processor can consume a large amount of memory and time during parsing. This may result in a DoS attack.

In the following example, when the `XMLConstants.FEATURE_SECURE_PROCESSING` feature is set to `false`, an attacker may perform an XML [billion laughs attack](https://en.wikipedia.org/wiki/Billion_laughs_attack):

```xml
```